core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -184,10 +184,44 @@ class SnappySessionState(snappySession: SnappySession)
     }
   }
 
-  case class AnalyzeChildQuery(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+  case class AnalyzeMutableOperations(sparkSession: SparkSession,
+      analyzer: Analyzer) extends Rule[LogicalPlan] {
+
+    private def getKeyAttributes(table: LogicalPlan,
+        child: LogicalPlan, plan: LogicalPlan): Seq[NamedExpression] = {
+      val keyColumns = table.collectFirst {
+        case LogicalRelation(mutable: MutableRelation, _, _) =>
+          val ks = mutable.getKeyColumns
+          if (ks.isEmpty) throw new AnalysisException(
+            s"Empty key columns for update/delete on $mutable")
+          ks
+      }.getOrElse(throw new AnalysisException(
```

Comment:
"Key" means a unique way to quickly get to the required rows.

For row tables it will be either the primary key, or the internal generated long (a ROWID) -- essentially the corresponding region key, or a "RowLocation" if you may. We cannot be having a case where every row update/delete leads to another full table scan. The idea being that during the corresponding query the rows have already been located and there is a unique handle to the row which delete/update can use to quickly get to the row. The current "deleteFrom" implementation suffers from this weakness and will be unusable when table has no primary key.

Likewise for column tables, the "key" will be the batchUUID+position to uniquely locate the row to be updated/deleted. In case there is no exchange/broadcast in the query portion, then this can be further optimized to directly pick up the handle of the DeltaColumn being currently scanned.